OrbitSsh/include/OrbitSsh/Task.h

```diff
@@ -0,0 +1,134 @@
+  std::optional<ResultType> result_ = std::nullopt;
+  Session* session_ptr_;
+  std::optional<Channel> channel_;
+  std::string command_;
+  std::function<void(std::string, std::optional<ResultType>*)>
+      std_out_callback_;
+  std::function<void(std::string, std::optional<ResultType>*)>
+      std_err_callback_;
+  std::function<ResultType(int, std::optional<ResultType>*)> exit_callback_;
+  State state_ = State::kNotInitialized;
+};
+
+// template implementation
+template <typename ResultType>
+Task<ResultType>::Task(
+    Session* session_ptr, std::string command,
+    std::function<void(std::string, std::optional<ResultType>*)>
+        std_out_callback,
+    std::function<void(std::string, std::optional<ResultType>*)>
+        std_err_callback,
+    std::function<ResultType(int, std::optional<ResultType>*)> exit_callback)
+    : session_ptr_(session_ptr),
+      command_(command),
+      std_out_callback_(std_out_callback),
+      std_err_callback_(std_err_callback),
+      exit_callback_(exit_callback) {
+  FAIL_IF(!exit_callback_, "Did not provide valid exit callback");
+}
+
+template <typename ResultType>
+outcome::result<void> Task<ResultType>::Read() {
+  if (std_out_callback_) {
+    auto std_out = channel_->Read();
+
+    // if std out reading has error -> return it
+    if (!std_out && !shouldITryAgain(std_out)) return std_out.error();
+
+    // if std out not empty -> std out callback
+    if (std_out && !std_out.value().empty()) {
+      std_out_callback_(std_out.value(), &result_);
+    }
+  }
+
+  if (std_err_callback_) {
+    auto std_err = channel_->ReadStdErr();
+
+    // if std err reading has error -> return it
+    if (!std_err && !shouldITryAgain(std_err)) return std_err.error();
+
+    // if std err not empty -> std err callback
+    if (std_err && !std_err.value().empty()) {
+      std_err_callback_(std_err.value(), &result_);
+    }
+  }
+
+  return outcome::success();
+}
+
+// Tick progresses the state when appropriate and is responsible for calling
+// std_out_callback and exit_callback. This function should be called
+// periodically.
+template <typename ResultType>
+outcome::result<ResultType> Task<ResultType>::Run() {
+  switch (state_) {
+    case State::kNotInitialized: {
+      OUTCOME_TRY(channel, Channel::OpenChannel(session_ptr_));
+      channel_ = std::move(channel);
+      state_ = State::kChannelOpened;
+    }
+    case State::kChannelOpened: {
+      OUTCOME_TRY(channel_->Exec(command_));
+      state_ = State::kRunning;
+    }
+    case State::kRunning: {
+      OUTCOME_TRY(Read());
+
+      // when remote did not sent EOF -> there will be more data in the future
+      if (!channel_->GetRemoteEOF()) return Error::kEagain;
+
+      state_ = State::kFinished;
+      return outcome::success(
+          exit_callback_(channel_->GetExitStatus(), &result_));
+    }
+    case State::kFinished:
+      return Error::kTaskUsedAfterFinish;
+      break;
+  }
+
+  return outcome::success();
```

Comment:
This makes no sense anymore, right? Error::kTaskUsedAfterFinish should trigger here, right?

Reply:
true, done